Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModule.cs

```diff
@@ -1490,25 +1567,48 @@ private void OnPoint(InputAction.CallbackContext context)
 
         private void OnLeftClick(InputAction.CallbackContext context)
         {
+            if (CheckForRemovedDevice(ref context))
+                return;
+
             ref var state = ref GetPointerStateFor(ref context);
             state.leftButton.isPressed = context.ReadValueAsButton();
             state.changedThisFrame = true;
         }
 
         private void OnRightClick(InputAction.CallbackContext context)
         {
+            if (CheckForRemovedDevice(ref context))
+                return;
+
             ref var state = ref GetPointerStateFor(ref context);
             state.rightButton.isPressed = context.ReadValueAsButton();
             state.changedThisFrame = true;
         }
 
         private void OnMiddleClick(InputAction.CallbackContext context)
         {
+            if (CheckForRemovedDevice(ref context))
+                return;
+
             ref var state = ref GetPointerStateFor(ref context);
             state.middleButton.isPressed = context.ReadValueAsButton();
             state.changedThisFrame = true;
         }
 
+        private bool CheckForRemovedDevice(ref InputAction.CallbackContext context)
+        {
+            // When a device is removed, we want to simply cancels ongoing pointer
```

Comment:
(A question) Would it be possible for the native-backend/runtime/input-manager-event-consumption-logic to guarantee if device is disconnected then we always have an event to "depress"/"cancel"/"release" touch-mouse-etc controls? I'm concerned that spreading this logic all over the code would create challenges in the future, for example would composite input actions correctly transition through the phases if I randomly disconnect mouse-like devices when they are in performed state?

Reply:
>Would it be possible for the native-backend/runtime/input-manager-event-consumption-logic to guarantee if device is disconnected then we always have an event to "depress"/"cancel"/"release" touch-mouse-etc controls?

In general, this is already happening. When a device is removed, its actions get cancelled (see `Actions_WhenDeviceIsRemoved_OngoingActionsAreCancelled`). So, usually you don't have to do this thing here.

The reason it's done in this case is due to the `ReadValue` on the touch ID control we're doing in `GetPointerStateIndexFor`.

Lemme walk through this all the way from the beginning... (warning: going to get messy....). May be useful for arriving at a better setup.

So, the UI module has a couple of actions sourcing input. However, while these are all separate actions, they are not really independent. `point`, `leftClick`, `scrollWheel`, etc. all go together. As does `navigate`, `submit`, and `cancel`. In other words, these form compound inputs.

Unfortunately, we can't use composites (at least for the pointer-type stuff) as we are sourcing arbitrary many pointers through these actions.

So... when we get input on the `point` action, we somehow need to associate that with a specific pointer in the UI that should receive the position value coming in.

That's where `touchId` is entering the picture. Normally, we simply say "one device, one pointer". So we can just use the device ID as a pointer  ID. But for touches, there may be several pointers coming in from the same device. So we need something more complicated there. And ATM that's handled by `GetPointerStateIndexFor` which reads touch IDs explicitly, if necessary.

However.... that means we have input that we read in the UI module which is *not* coming in through actions. Which is... bad.

And... which leads us back to the beginning. The reason that the UI module, unlike other parts of the code, explicitly is checking for device removals is to prevent itself from reading touch IDs when it shouldn't.

Reply:
Got it.